### PR TITLE
Add eclipse support

### PR DESCRIPTION
**Pull request by rougsig**
More details here https://github.com/google/protobuf-gradle-plugin/pull/596

### CODE REVIEW

**Comment by rougsig** [comment]
Wait for https://github.com/google/protobuf-gradle-plugin/pull/599

**Comment by rougsig** [comment]
I leave IDESupportTest with fail. It's expected behaviour for now.

Open questions:
1. When we set up eclipse with the eclipse plugin, we can add any attributes. With the Gradle vanilla API, we can't do that.
Is it necessary to pass all the attributes that were in the test?

2. The :eclipse task only includes existing folders in the .classpath. That is, this folder is registered as a sourceSet, but if the folder does not exist on disk, it will not be included in the .classpath file.
Is it okay?

cc @jdneo @ejona86

**Comment by jdneo** [comment]
> When we set up eclipse with the eclipse plugin, we can add any attributes. With the Gradle vanilla API, we can't do that.

Does Gradle vanilla API support a limit set of the attributes? Or no attributes are supported at all?

> if the folder does not exist on disk, it will not be included in the .classpath file.

This is the purpose of the attributes `optional` & `ignore_optional_problems`. An optional classpath entry is allowed to be added into `.classpath` even it does not exist on the disk in Eclipse.

If the folder does not included in the .classpath file, after user compiles the proto files to Java sources, user needs to reload the project to update the .classpath file, otherwise, there will still be a lot of compilation errors displayed. That might not be ideal as it's not out-of-box experience - user might not be aware that such action is needed to make it work.

**Comment by rougsig** [comment]
> Does Gradle vanilla API support a limit set of the attributes? Or no attributes are supported at all?

As i know - not supported to add any attributes. All required attributes will be set automatically.

> That might not be ideal as it's not out-of-box experience - user might not be aware that such action is needed to make it work.

You are right.

Look like we need to keep custom logic for eclipse and idea.

---
@jdneo 
I'm not familiar with eclipse. I used it a long time ago when the idea didn't exist yet. When does eclipse generate a .classpath file? Is it generated automatically or does the user need to execute a specific command?

**Comment by jdneo** [comment]
> When does eclipse generate a .classpath file? Is it generated automatically or does the user need to execute a specific command?

It's generated automatically when user is importing a Gradle project in Eclipse. There is a plug-in in Eclipse called [Buildship](https://github.com/eclipse/buildship), this plugin is used to parse the Gradle project to an Eclipse project (just like when we import a Maven or Gradle project into IDEA, IDEA will try to understand the project and parse it into IDEA's own model).

The .classpath file is generated when Buildship is doing the import stuff.

**Comment by rougsig** [comment]
I moved the eclipse fix to this PR for the reason that registering the sourceset should fix the issue. Sourceset registration actually solves the problem, but not in the best way for the user experience. IDEA works well without custom code from the gradle idea plugin. Eclipse doesn't work perfectly without custom code from eclipse gradle plugin. I think the best solution will be keep custom code for eclipse. 

@jdneo How do you feel better, will you merge https://github.com/google/protobuf-gradle-plugin/pull/596 or will we leave these changes in this PR?

**Comment by jdneo** [comment]
> I think the best solution will be keep custom code for eclipse.

Agree.

> will you merge https://github.com/google/protobuf-gradle-plugin/pull/596 or will we leave these changes in this PR?

I can append a commit to this PR (Not sure if I have the push permission to your fork). Or you can merge this one and I can update #596 based on the latest change.

**Comment by jdneo** [comment]
Hi @rougsig 👋,

I found I have no permission to append commit to your fork. So, I updated #596. (Two deprecated attributes are removed)

**Comment by rougsig** [comment]
I keep for now idea custom code. 

@jdneo I fix lint problems, and move IDE support tests to separate file. All other changes yours. Remove draft status.

**Comment by rougsig** [comment]
@jdneo Could you review this PR?

**Comment by jdneo** [comment]
Thank you @rougsig, I'll review it today.

**Review by jdneo — APPROVED** [review]
LGTM!

Now Eclipse and VS Code can automatically add the protobuf output source directories into the source paths, without manually & explicitly set the source set in `build.gradle`.

##### Eclipse
protobuf-eclipse

##### VS Code
protobuf-vscode

File: src/main/groovy/com/google/protobuf/gradle/Utils.groovy [hunk 1]
```diff
@@ -129,4 +133,78 @@ class Utils {
...
+        String relativePath = projectDir.toURI().relativize(f.toURI()).getPath()
+        // remove trailing slash
+        if (relativePath.endsWith("/")) {
+          relativePath = relativePath[0..(relativePath.length() - 2)]
+        }
+        SourceFolder entry = new SourceFolder(relativePath, getOutputPath(cp, sourceSetName))
+        entry.entryAttributes.put("optional", "true")
+        entry.entryAttributes.put("ignore_optional_problems", "true")
+
+        // this attribute is optional, but it could be useful for IDEs to recognize that it is
```
Comment (ejona86) [inline src/main/groovy/com/google/protobuf/gradle/Utils.groovy:160]: I'm wary of adding additional information like this. Ideally we wouldn't use an Eclipse API, and we can't have this sort of extra information in a generic API. So having it here leads to one of two possible futures: 1) it is unused and doesn't matter or 2) it is used and now the plugin must always use the Eclipse API.

File: src/main/groovy/com/google/protobuf/gradle/Utils.groovy [hunk 2]
```diff
@@ -129,4 +133,78 @@ class Utils {
...
+      }
+    }
+  }
+
+  /**
+   * Get the output path according to the source set name, if no valid path can be
+   * found, <code>null</code> will return.
+   */
+  private static String getOutputPath(Classpath classpath, String sourceSetName) {
+    String path = "bin/" + sourceSetName
```
Comment (ejona86) [inline src/main/groovy/com/google/protobuf/gradle/Utils.groovy:180]: This is the output path of what? The compiled classes output? I don't see it asserted in the test either.

File: src/test/groovy/com/google/protobuf/gradle/IDESupportTest.groovy [hunk 3]
```diff
@@ -0,0 +1,142 @@
+package com.google.protobuf.gradle
```
Comment (ejona86) [inline src/test/groovy/com/google/protobuf/gradle/IDESupportTest.groovy:1]: Add the copyright header.

**Review by ejona86 — COMMENTED** [review]
> Sourceset registration actually solves the problem, but not in the best way for the user experience. ... Eclipse doesn't work perfectly without custom code from eclipse gradle plugin.

What is wrong with Eclipse support when we don't have the custom code? What is wrong with having code that just creates the directories, and nothing else?

File: src/main/groovy/com/google/protobuf/gradle/Utils.groovy (hunk 2 above)
Comment (jdneo) [reply src/main/groovy/com/google/protobuf/gradle/Utils.groovy:180]: Yes it's the compiled classes output.

Buildship(Eclipse Gradle project importer) by default will set the output folder to `bin`

**Review by ejona86 — APPROVED** [review]

**Pull request merged**